File: Php/app/Cashgame/Action/ActionChartData.php
namespace app\Cashgame\Action {

	use DateTime;
	use app\Chart\ChartDateTimeColumnModel;
	use app\Chart\ChartDateTimeValueModel;
	use app\Chart\ChartModel;
	use app\Chart\ChartNumberColumnModel;
	use app\Chart\ChartRowModel;
	use app\Chart\ChartValueModel;
	use core\DateTimeFactory;
	use entities\Cashgame;
	use entities\CashgameResult;
	use entities\Checkpoints\CheckpointType;
	use entities\GameStatus;
	use entities\Homegame;

	class ActionChartData extends ChartModel{

		private $homegame;
		private $cashgame;
		private $result;

		public function __construct(Homegame $homegame, Cashgame $cashgame, CashgameResult $result){
			homegame = $homegame;
			cashgame = $cashgame;
			result = $result;
			addActionColumns();
			addActionRows();
		}

		private function addActionRows(){
			$checkpoints = getCheckpoints();
			$totalBuyin = 0;
			foreach($checkpoints as $checkpoint){
				if($checkpoint.getType() == CheckpointType::buyin){
					if($totalBuyin > 0){
						$stackBefore = $checkpoint.getStack() - $checkpoint.getAmount();
						addActionRow($checkpoint.getTimestamp(), $stackBefore, $totalBuyin);
					}
					$totalBuyin += $checkpoint.getAmount();
				}
				addActionRow($checkpoint.getTimestamp(), $checkpoint.getStack(), $totalBuyin);
			}
			if(cashgame.getStatus() == GameStatus::running){
				$timestamp = DateTimeFactory::now(homegame.getTimezone());
				addActionRow($timestamp, result.getStack(), result.getBuyin());
			}
		}

		private function getCheckpoints(){
			if(playerIsInGame()){
				return result.getCheckpoints();
			} else {
				return array();
			}
		}

		private function playerIsInGame(){
			return result != null;
		}

		private function addActionColumns(){
			$timeCol = new ChartDateTimeColumnModel('Time', 'HH:mm');
			$stackCol = new ChartNumberColumnModel('Stack');
			$buyinCol = new ChartNumberColumnModel('Buyin');
			addColumn($timeCol);
			addColumn($stackCol);
			addColumn($buyinCol);
		}

		private function addActionRow(DateTime $dateTime, $stack, $buyin){
			$row1 = new ChartRowModel();
			$row1.addValue(new ChartDateTimeValueModel($dateTime));
			$row1.addValue(new ChartValueModel($stack));
			$row1.addValue(new ChartValueModel($buyin));
			addRow($row1);
		}

	}

}

File: Web/Models/CashgameModels/Action/ActionChartData.cs
using System;
using System.Collections.Generic;
using Core.Classes;
using Core.Classes.Checkpoints;
using Infrastructure.System;
using Web.Models.ChartModels;

namespace Web.Models.CashgameModels.Action {

	public class ActionChartData : ChartModel{

		public ActionChartData(Homegame homegame, Cashgame cashgame, CashgameResult result){
			AddActionColumns();
			AddActionRows(homegame, cashgame, result);
		}

        private void AddActionRows(Homegame homegame, Cashgame cashgame, CashgameResult result)
        {
			var checkpoints = GetCheckpoints(result);
			var totalBuyin = 0;
			foreach(var checkpoint in checkpoints){
				if(checkpoint.Type == CheckpointType.Buyin){
					if(totalBuyin > 0){
						var stackBefore = checkpoint.Stack - checkpoint.Amount;
						AddActionRow(checkpoint.Timestamp, stackBefore, totalBuyin);
					}
					totalBuyin += checkpoint.Amount;
				}
				AddActionRow(checkpoint.Timestamp, checkpoint.Stack, totalBuyin);
			}
			if(cashgame.Status == GameStatus.Running){
				var timestamp = DateTimeFactory.Now(homegame.Timezone);
				AddActionRow(timestamp, result.Stack, result.Buyin);
			}
		}

		private IEnumerable<Checkpoint> GetCheckpoints(CashgameResult result){
			if(PlayerIsInGame(result)){
				return result.Checkpoints;
			} else {
				return new List<Checkpoint>();
			}
		}

		private bool PlayerIsInGame(CashgameResult result){
			return result != null;
		}

		private void AddActionColumns(){
			var timeCol = new ChartDateTimeColumnModel("Time", "HH:mm");
			var stackCol = new ChartNumberColumnModel("Stack");
			var buyinCol = new ChartNumberColumnModel("Buyin");
			AddColumn(timeCol);
			AddColumn(stackCol);
			AddColumn(buyinCol);
		}

		private void AddActionRow(DateTime dateTime, int stack, int buyin){
			var row = new ChartRowModel();
			row.AddValue(new ChartDateTimeValueModel(dateTime));
			row.AddValue(new ChartValueModel(stack));
			row.AddValue(new ChartValueModel(buyin));
			AddRow(row);
		}

	}

}